Allow adding custom annotations to pods

amit-tewari opened the pull request:
for use case such as metrics scraping annotations

## Why are the changes needed
Current chart templates allow enabling metrics export, but do not allow setting annotations so Prometheus (or similar tools) can run discovery and scrape metrics.

This PR allows possibility to add custom annotations.

patch version bump is done, since this change does not break existing usage, but allows opt-in to add new annotations.

## How to test this PR
on master branch
```
helm template my-presto charts/presto --set ingress.enabled=true --debug > /tmp/old.yaml
```
on PR branch
```
helm template my-presto charts/presto --set ingress.enabled=true --debug > /tmp/new.yaml
```
finally see changes
```
 diff -y /tmp/old.yaml /tmp/new.yaml  | less
```

@dnskr ,
Please have a look, and let me know if any adjustments are needed.

amit-tewari committed:
Allow adding custom annotations to pods

for use case such as metrics scraping annotations

dnskr commented on charts/presto/Chart.yaml:
```diff
@@ -2,7 +2,7 @@ apiVersion: v2
 description: The official Helm chart for Presto
 type: application
-version: 0.4.0
+version: 0.4.1
```
```suggestion
version: 0.4.0
```
The proposed change is not a bug fix and should be released in the next minor version of the chart. Therefore, please do not update the chart version, as that would trigger the release workflow.

amit-tewari replied:
done.

dnskr commented on charts/presto/templates/deployment-coordinator.yaml:
```diff
@@ -37,6 +37,9 @@ spec:
       annotations:
         checksum/catalog: {{ include (print $.Template.BasePath "/configmap-catalog.yaml") . | sha256sum }}
         checksum/coordinator: {{ include (print $.Template.BasePath "/configmap-coordinator.yaml") . | sha256sum }}
+        {{- with .Values.coordinator.podAnnotations }}
+        {{- toYaml . | nindent 8 }}
+        {{- end }}
```
```suggestion
        {{- with .Values.coordinator.podAnnotations }}
          {{- toYaml . | nindent 8 }}
        {{- end }}
```
Could you please align the indentation to follow the common style (similar to [service labels](https://github.com/prestodb/presto-helm-charts/blob/5aa7ad008578fec10915515bf76a6d868129a322/charts/presto/templates/service.yaml#L6-L13))?

amit-tewari replied:
done.

dnskr reviewed (changes requested):
Thank you for submitting the PR!

Having `podAnnotations` support seems to be a must-have in the chart.
What do you think about adding a common `podAnnotations` property as well, and appending coordinator/resource manager/worker annotations similarly to how it's done with `env`, `envFrom`, `initContainers`, etc.?

amit-tewari committed:
incorporate review comments

amit-tewari commented:
accommodated review inputs.

dnskr commented on charts/presto/templates/deployment-coordinator.yaml:
```diff
@@ -1,3 +1,4 @@
+{{- $podAnnotations := mergeOverwrite .Values.coordinator.podAnnotations .Values.podAnnotations }}
```
```suggestion
{{- $podAnnotations := mergeOverwrite .Values.podAnnotations .Values.coordinator.podAnnotations }}
```
The `mergeOverwrite` function merges dictionaries into one, giving precedence from right to left, see [docs](https://helm.sh/docs/chart_template_guide/function_list/#mergeoverwrite-mustmergeoverwrite).

I think that the value of a specific common annotation should be overwritten by the component-specific value.
For instance, the following values:
```yaml
podAnnotations:
  annotation-source-common: "true"
  check-merge-overwrite: "common"
coordinator:
  podAnnotations:
    annotation-source-coordinator: "true"
    check-merge-overwrite: "coordinator"
```
should render `coordinator` for the `check-merge-overwrite` value instead of `common`, i.e.
```yaml

spec:
  template:
    metadata:
      annotations:
        ...
        annotation-source-common: "true"
        annotation-source-coordinator: "true"
        check-merge-overwrite: coordinator
```

amit-tewari replied:
this is fixed now.

amit-tewari committed:
fix annotations merge ordering

dnskr commented on charts/presto/templates/deployment-coordinator.yaml:
```diff
@@ -1,3 +1,4 @@
+{{- $podAnnotations := mergeOverwrite .Values.podAnnotations .Values.coordinator.podAnnotations }}
```
I found that, unfortunately, it doesn't work as intended.

The following command:
```shell
helm template my-presto charts/presto --set mode=ha-cluster
```

and `values.yaml`:
```yaml
# Annotations to add to each pod in every deployment
podAnnotations:
  annotation-source-common: "true"
  check-merge-overwrite: "common"

resourceManager:
  # Annotations to add to resource manager pods
  podAnnotations:
    annotation-source-resource-manager: "true"
    check-merge-overwrite: "resource-manager"

coordinator:
  # Annotations to add to coordinator pods
  podAnnotations:
    annotation-source-coordinator: "true"
    check-merge-overwrite: "coordinator"

worker:
  # Annotations to add to worker pods
  podAnnotations:
    annotation-source-worker: "true"
    check-merge-overwrite: "worker"
```

give incorrect templating result:
```yaml
# Source: presto/templates/deployment-coordinator.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  ...
spec:
  ...
  template:
    metadata:
      ...
      annotations:
        checksum/catalog: 45dd1075e7d7116412d8ccd22dd6095e927cf4aaaa6d13f3d3f97047f90cac37
        checksum/coordinator: fcbb58e8ed85d70247e1fa231cc8c9e806ea15ee9bb11b3ca615a78aead9516d
        annotation-source-common: "true"
        annotation-source-coordinator: "true"
        annotation-source-resource-manager: "true"
        annotation-source-worker: "true"
        check-merge-overwrite: coordinator
```

---
The [mergeOverwrite](https://helm.sh/docs/chart_template_guide/function_list/#mergeoverwrite-mustmergeoverwrite) function merges dictionaries into one, overwriting values in the destination dictionary. The documentation recommends using the `deepCopy` function with merging.

amit-tewari replied:
issue is fixed in next comment.

amit-tewari committed:
Fix for issue detected in review comment

Unfortunately none of the function works in direct template. It seems
all variables are computed across all templates before rendering starts.

To fix this, helper funtions are created that exhibit expected
behaviour.

Related comment:
https://github.com/prestodb/presto-helm-charts/pull/7/files/70dc748c2144d1772d74bf884240f5357950ec80#r2083639010

dnskr commented:
What do you think about using `mergeOverwrite` with `deepCopy` to avoid overwriting common annotations?
```yaml
{{- $podAnnotations := mergeOverwrite (deepCopy .Values.podAnnotations) .Values.coordinator.podAnnotations }}
...
apiVersion: apps/v1
kind: Deployment
metadata:
  ...
spec:
  ...
  template:
    metadata:
      labels:
        ...
      annotations:
        checksum/catalog: {{ include (print $.Template.BasePath "/configmap-catalog.yaml") . | sha256sum }}
        checksum/coordinator: {{ include (print $.Template.BasePath "/configmap-coordinator.yaml") . | sha256sum }}
        {{- with $podAnnotations }}
          {{- toYaml . | nindent 8 }}
        {{- end }}
```

Example in [airflow/chart/templates/workers/worker-deployment.yaml](https://github.com/apache/airflow/blob/35a4ced38b5c45dc9f438e41e6ac1a6bff1a5692/chart/templates/workers/worker-deployment.yaml#L42).

amit-tewari commented:
> What do you think about using `mergeOverwrite` with `deepCopy` to avoid overwriting common annotations?
> 
> ```yaml
> {{- $podAnnotations := mergeOverwrite (deepCopy .Values.podAnnotations) .Values.coordinator.podAnnotations }}
> ...
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   ...
> spec:
>   ...
>   template:
>     metadata:
>       labels:
>         ...
>       annotations:
>         checksum/catalog: {{ include (print $.Template.BasePath "/configmap-catalog.yaml") . | sha256sum }}
>         checksum/coordinator: {{ include (print $.Template.BasePath "/configmap-coordinator.yaml") . | sha256sum }}
>         {{- with $podAnnotations }}
>           {{- toYaml . | nindent 8 }}
>         {{- end }}
> ```
> 
> Example in [airflow/chart/templates/workers/worker-deployment.yaml](https://github.com/apache/airflow/blob/35a4ced38b5c45dc9f438e41e6ac1a6bff1a5692/chart/templates/workers/worker-deployment.yaml#L42).

I guess it depends on what behavior we intend to have.

1. similar to `/etc/` config values overridden by `$XDG_CONFIG_HOME` config values, OR
2. other way round,  `$XDG_CONFIG_HOME` config values overridden by `/etc` config values

airflow worker seems to follow preference 2, where as this MR follows along the lines of 1. I personally have no preference, since both approaches have their own pros and cons. I respect your choice as the maintainer of the chart to be in the best interest of the community. Please let me know your preference, and I can adjust the config the way you propose (also, I would not mind if you make changes to this MR yourself, to conclude faster... I trust you completely :pray: ).

dnskr committed:
Revert "Fix for issue detected in review comment"

This reverts commit 54dbe9a74887f0143d54a64a88b31e197854076e.

dnskr committed:
Use deepCopy to not mutate podAnnotations dictionary

dnskr commented:
@amit-tewari I reverted the latest commit and added a minor fix, so it now works as expected.

The following command:
```shell
helm template my-presto charts/presto --set mode=ha-cluster
```

and `values.yaml`:
```yaml
# Annotations to add to each pod in every deployment
podAnnotations:
  annotation-source-common: "true"
  check-merge-overwrite: "common"

resourceManager:
  # Annotations to add to resource manager pods
  podAnnotations:
    annotation-source-resource-manager: "true"
    check-merge-overwrite: "resource-manager"

coordinator:
  # Annotations to add to coordinator pods
  podAnnotations:
    annotation-source-coordinator: "true"
    check-merge-overwrite: "coordinator"

worker:
  # Annotations to add to worker pods
  podAnnotations:
    annotation-source-worker: "true"
    check-merge-overwrite: "worker"
```

give correct results:
```yaml
---
# Source: presto/templates/deployment-coordinator.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-presto-coordinator
  ...
spec:
  ...
  template:
    metadata:
      ...
      annotations:
        checksum/catalog: 45dd1075e7d7116412d8ccd22dd6095e927cf4aaaa6d13f3d3f97047f90cac37
        checksum/coordinator: fcbb58e8ed85d70247e1fa231cc8c9e806ea15ee9bb11b3ca615a78aead9516d
        annotation-source-common: "true"
        annotation-source-coordinator: "true"
        check-merge-overwrite: coordinator

---
# Source: presto/templates/deployment-resource-manager.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-presto-resource-manager
  ...
spec:
  ...
  template:
    metadata:
      ...
      annotations:
        checksum/catalog: 45dd1075e7d7116412d8ccd22dd6095e927cf4aaaa6d13f3d3f97047f90cac37
        checksum/resource-manager: 83116030cae7281139e85ab3a370a8550aab6cc917bc629622ec14d67f6d76eb
        annotation-source-common: "true"
        annotation-source-resource-manager: "true"
        check-merge-overwrite: resource-manager

---
# Source: presto/templates/deployment-worker.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-presto-worker
  ...
spec:
  ...
  template:
    metadata:
      ...
      annotations:
        checksum/catalog: 45dd1075e7d7116412d8ccd22dd6095e927cf4aaaa6d13f3d3f97047f90cac37
        checksum/worker: a7ea8f869711764be8a165fefa7b4a22236664316c1b487cb6fc97afa570e621
        annotation-source-common: "true"
        annotation-source-worker: "true"
        check-merge-overwrite: worker
```

---
Could you please have a look and confirm if it makes sense?

amit-tewari commented:
Thank you. :pray: 
This looks good.

dnskr commented:
Thank you @amit-tewari for contributing and being patient with a long review process!

amit-tewari commented:
Thank you for collaboration. :pray:
